### PR TITLE
Issue #880 Merge Cache implementations

### PR DESCRIPTION
**Pull request by ljacomet**
* Merging back `Cache` implementations in a single `Ehcache` class
* Implement `Ehcache.get` short circuit that invokes `Store.get`

### CODE REVIEW

File: core-spi-test/src/main/java/org/ehcache/internal/store/StoreReplaceKeyValueValueTest.java [hunk 1]
```diff
@@ -151,7 +151,7 @@ public void wrongKeyTypeThrowsException()
 
     try {
       if (this.factory.getKeyType() == String.class) {
```
Comment (chrisdennis) [inline core-spi-test/src/main/java/org/ehcache/internal/store/StoreReplaceKeyValueValueTest.java]: Not your code, but this type check is wrong, it should be: `this.factory.getKeyType().isAssignableFrom(String.class)`

File: impl/src/test/java/org/ehcache/impl/config/serializer/SerializerCountingTest.java [hunk 2]
```diff
@@ -92,7 +91,7 @@ public void testOnHeapPutGet() {
     printSerializationCounters("Get OnHeap");
 
     cache.put(42L, "Wrong ...");
-    assertCounters(2, 2, 0, 1, 0, 0);
+    assertCounters(2, 2, 0, 1, 1, 0);
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/impl/config/serializer/SerializerCountingTest.java]: The numbers went up... do we care about that?

Comment (ljacomet) [inline impl/src/test/java/org/ehcache/impl/config/serializer/SerializerCountingTest.java]: Put that underneath uses compute does indeed add extra deserialization of value. Adding a general comment to attract attention.

**Comment by ljacomet** [comment]
This change introduces extra deserialization on put - see `SerializerCountingTest`. @lorban what was the perf status for put?

**Comment by ljacomet** [comment]
Not doing for now given clustering discussion with @chrisdennis, @lorban and @cljohnso

**Pull request closed without merging**